Local directory cache can fail on race condition

Some builds can throw exceptions like this while loading from the local cache:

```text
Failed to load cache entry for task ':lib:twitter:analytics:service:generateDebugResValues', falling back to executing task
org.gradle.api.GradleException: Build cache entry 1907a3b377f942b554a1e8a24ab646fb2758846954718045699 from local build cache is invalid
Caused by: org.gradle.api.UncheckedIOException: java.io.FileNotFoundException: /Users/lptr/Workspace/gradle/gradle/build/tmp/test files/CachedCustomTaskExecutionIntegrationTest/concurrent_access_t...che_works/ezrg8/cache-dir/1907a3b377f942b554a1e8a24ab646fb (No such file or directory)
```

Another variant is an error encountered while storing to the local cache:

```text
Failed to store cache entry 1907a3b377f942b554a1e8a24ab646fb for task ':r2:test'
org.gradle.internal.resource.local.FileStoreException: Failed to move file '/Users/lptr/Workspace/gradle/gradle/build/tmp/test files/CachedCustomTaskExecutionIntegrationTest/concurrent_access_t...che_works/ezrg8/cache-dir/1907a3b377f942b554a1e8a24ab646fb2758846954718045699.part' into filestore at '/Users/lptr/Workspace/gradle/gradle/build/tmp/test files/CachedCustomTaskExecutionIntegrationTest/concurrent_access_t...che_works/ezrg8/cache-dir/1907a3b377f942b554a1e8a24ab646fb' 
```

To reproduce, we need at least three instances of the same task in different projects (so that they can be executed in parallel) with the exact same inputs (so that their cache key ends up being the same). Then the tasks need to be executed in roughly this order:

```
Task 1: [CHECK: no] [EXECUTE] [STORE]
Task 2:     [CHECK: no] [EXECUTE--------------] [STORE]
Task 3:                            [CHECK: yes] [LOAD]
```

`Task 1` first checks whether there is a cached result with the cache key, and because there's none, it ends up executing. Once it has finished executing its output is stored in the cache.

`Task 2` does the same thing: when it first checks the local cache for a result, it finds none, because `Task 1` hasn't finished and stored yet. So `Task 2` too starts executing, and gets stored, too.

The code in Gradle works as such that the second `[STORE]` operation first deletes the already stored output in the cache. If a third task (`Task 3`) tries to `[LOAD]` from the cache while this is happening, the fist exception above is thrown.

When multiple `[STORE]` operations align just so, then the second exception is produced.

--- a/subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/DefaultBuildCacheController.java
+++ b/subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/DefaultBuildCacheController.java
@@ -282,11 +282,7 @@ public class DefaultBuildCacheController implements BuildCacheController {
     }
 
     private static LocalBuildCacheServiceHandle toHandle(LocalBuildCacheService local, boolean localPush) {
-        if (local == null) {
-            return NullLocalBuildCacheServiceHandle.INSTANCE;
-        } else {
-            return new DefaultLocalBuildCacheServiceHandle(local, localPush);
-        }
+        return new DefaultLocalBuildCacheServiceHandle(local, localPush);
     }
 
 }

--- a/subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheService.java
+++ b/subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheService.java
@@ -30,10 +30,15 @@ import org.gradle.internal.resource.local.LocallyAvailableResource;
 import org.gradle.internal.resource.local.PathKeyFileStore;
 import org.gradle.util.GFileUtils;
 
+import javax.annotation.Nonnull;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.attribute.FileTime;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 public class DirectoryBuildCacheService implements LocalBuildCacheService, BuildCacheService {
 
@@ -41,6 +46,7 @@ public class DirectoryBuildCacheService implements LocalBuildCacheService, Build
     private final PersistentCache persistentCache;
     private final BuildCacheTempFileStore tempFileStore;
     private final String failedFileSuffix;
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
 
     public DirectoryBuildCacheService(PathKeyFileStore fileStore, PersistentCache persistentCache, BuildCacheTempFileStore tempFileStore, String failedFileSuffix) {
         this.fileStore = fileStore;
@@ -58,7 +64,7 @@ public class DirectoryBuildCacheService implements LocalBuildCacheService, Build
         }
 
         @Override
-        public void execute(File file) {
+        public void execute(@Nonnull File file) {
             try {
                 // Mark as recently used
                 GFileUtils.touch(file);
@@ -86,37 +92,48 @@ public class DirectoryBuildCacheService implements LocalBuildCacheService, Build
 
     @Override
     public void load(final BuildCacheKey key, final Action<? super File> reader) {
-        // We need to lock here because garbage collection can be under way in another process
+        // We need to lock other processes out here because garbage collection can be under way in another process
         persistentCache.withFileLock(new Runnable() {
             @Override
             public void run() {
-                LocallyAvailableResource resource = fileStore.get(key.getHashCode());
-                if (resource != null) {
-                    final File file = resource.getFile();
-                    GFileUtils.touch(file); // Mark as recently used
-
-                    try {
-                        reader.execute(file);
-                    } catch (Exception e) {
-                        // Try to move the file out of the way in case its permanently corrupt
-                        // Don't delete, so that it can be potentially used for debugging
-                        File failedFile = new File(file.getAbsolutePath() + failedFileSuffix);
-                        GFileUtils.deleteQuietly(failedFile);
-                        //noinspection ResultOfMethodCallIgnored
-                        file.renameTo(failedFile);
-
-                        throw UncheckedException.throwAsUncheckedException(e);
-                    }
+                lock.readLock().lock();
+                try {
+                    loadInsideLock(key, reader);
+                } finally {
+                    lock.readLock().unlock();
                 }
             }
         });
+    }
+
+    private void loadInsideLock(BuildCacheKey key, Action<? super File> reader) {
+        LocallyAvailableResource resource = fileStore.get(key.getHashCode());
+        if (resource == null) {
+            return;
+        }
+
+        File file = resource.getFile();
+        touch(file);
+
+        try {
+            reader.execute(file);
+        } catch (Exception e) {
+            // Try to move the file out of the way in case its permanently corrupt
+            // Don't delete, so that it can be potentially used for debugging
+            File failedFile = new File(file.getAbsolutePath() + failedFileSuffix);
+            GFileUtils.deleteQuietly(failedFile);
+            //noinspection ResultOfMethodCallIgnored
+            file.renameTo(failedFile);
+
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
     }
 
     @Override
     public void store(final BuildCacheKey key, final BuildCacheEntryWriter result) throws BuildCacheException {
         tempFileStore.allocateTempFile(key, new Action<File>() {
             @Override
-            public void execute(final File file) {
+            public void execute(@Nonnull File file) {
                 try {
                     Closer closer = Closer.create();
                     try {
@@ -137,12 +154,21 @@ public class DirectoryBuildCacheService implements LocalBuildCacheService, Build
 
     @Override
     public void store(final BuildCacheKey key, final File file) {
-        persistentCache.useCache(new Runnable() {
+        persistentCache.withFileLock(new Runnable() {
             @Override
             public void run() {
-                fileStore.move(key.getHashCode(), file);
+                lock.writeLock().lock();
+                try {
+                    storeInsideLock(key, file);
+                } finally {
+                    lock.writeLock().unlock();
+                }
             }
         });
+    }
+
+    private void storeInsideLock(BuildCacheKey key, File file) {
+        fileStore.move(key.getHashCode(), file);
     }
 
     @Override
@@ -158,5 +184,14 @@ public class DirectoryBuildCacheService implements LocalBuildCacheService, Build
     @Override
     public void close() {
         persistentCache.close();
+    }
+
+    @SuppressWarnings("Since15")
+    private static void touch(File file) {
+        try {
+            Files.setLastModifiedTime(file.toPath(), FileTime.fromMillis(System.currentTimeMillis()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
     }
 }

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
@@ -18,7 +18,9 @@ package org.gradle.api.tasks
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
 import org.gradle.test.fixtures.file.TestFile
+import spock.lang.IgnoreIf
 import spock.lang.Issue
 import spock.lang.Unroll
 
@@ -773,6 +775,36 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
         true          | "{ null }"
         false         | "null"
         api = useRuntimeApi ? "runtime" : "annotation"
+    }
+
+    @IgnoreIf({GradleContextualExecuter.parallel})
+    @Issue("https://github.com/gradle/gradle/issues/3537")
+    def "concurrent access to local cache works"() {
+        def projectNames = GroovyCollections.combinations(('a'..'p'), ('a'..'p'), ('a'..'d'))*.join("")
+        println "Running with ${projectNames.size()} projects"
+        projectNames.each { projectName ->
+            settingsFile << "include '$projectName'\n"
+        }
+
+        buildFile << """
+            subprojects { project ->
+                task test {
+                    def outputFile = file("\${project.buildDir}/output.txt")
+                    outputs.cacheIf { true }
+                    outputs.file(outputFile).withPropertyName("outputFile") 
+                    doFirst {
+                        Thread.sleep(new Random().nextInt(100))
+                        outputFile.text = "output"
+                    }
+                }
+            }
+        """
+
+        when:
+        args "--parallel", "--max-workers=16"
+        withBuildCache().succeeds "test"
+        then:
+        noExceptionThrown()
     }
 
     private static String defineProducerTask() {

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
@@ -801,7 +801,7 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
         """
 
         when:
-        args "--parallel", "--max-workers=16"
+        args "--parallel", "--max-workers=40"
         withBuildCache().succeeds "test"
         then:
         noExceptionThrown()

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
@@ -793,7 +793,7 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
                     outputs.cacheIf { true }
                     outputs.file(outputFile).withPropertyName("outputFile") 
                     doFirst {
-                        Thread.sleep(new Random().nextInt(100))
+                        Thread.sleep(new Random().nextInt(30))
                         outputFile.text = "output"
                     }
                 }

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy
@@ -801,7 +801,7 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
         """
 
         when:
-        args "--parallel", "--max-workers=40"
+        args "--parallel", "--max-workers=100"
         withBuildCache().succeeds "test"
         then:
         noExceptionThrown()

